### ISSUE TITLE
Calculator to provide extrapolated "inputs"

### ISSUE THREAD

**Issue by GoFroggyRun** [open-source-economics/Tax-Calculator#507]
It seems that, if we were turning on the "stage one" and "stage two" extrapolations. Then the actual input file might not be useful, since what is being used that runs through the tax logic is the extrapolated input. And those values being used in calculation were hard to track.

Given my experience in dealing with #506, I think it would be helpful for our calculator to be able to produce a complete list of extrapolated values for each inputed variable, possibly as an alternative option. 

PS: I was able to get the extrapolated "inputs" manually, with the help of ipython-notebook, where the manual part involves modifying the uppercase and lowercase back and forth. Thus I think resolving the uppercase and lowercase struggles when recording certain inputs, as mentioned in #425, would be a first step.

**Comment by martinholmer** [comment]
Sean Wang ( @GoFroggyRun ) said:
> Given my experience in dealing with #506, I think it would be helpful for our calculator to be able to produce a complete list of extrapolated values for each inputed variable, possibly as an alternative option.
>
> PS: I was able to get the extrapolated "inputs" manually, with the help of ipython-notebook, where the manual part involves modifying the uppercase and lowercase back and forth. Thus I think resolving the uppercase and lowercase struggles when recording certain inputs, as mentioned in #425, would be a first step.

Two thoughts about this issue:

(1) I agree completely that the uppercase/lowercase differences in variable names between the input file ```puf.csv``` and the Python code are confusing and serve no purpose.  Eliminating these differences is part of the plan in resolving issue #425.

(2) I wonder if the debugging option you requested for the simtax.py and inctax.py programs would meet your needs for blown-up values of various input variables.  All you need to do is add the debugging variable names you want to add the output from simtax.py or inctax.py in the following SimpileTaxIO code:
```
    OVAR_NUM = 28
    DVAR_NAMES = [  # OPTIONAL DEBUGGING OUTPUT VARIABLE NAMES
        # '...',  # ovar[OVAR_NUM+1]
        # '...',  # ovar[OVAR_NUM+2]
        # etc.
        # '...'   # ovar[OVAR_NUM+n]
    ]
```
which then automatically activates the following logic:
```
        # add optional debugging output to ovar dictionary
        num = SimpleTaxIO.OVAR_NUM
        for dvar_name in SimpleTaxIO.DVAR_NAMES:
            num += 1
            dvar = getattr(crecs, dvar_name, None)
            if dvar is None:
                msg = 'debugging variable name "{}" not in calc.records object'
                raise ValueError(msg.format(dvar_name))
            else:
                ovar[num] = dvar[idx]
```
Does this approach meet your needs?  If not, can you explain exactly what variable values you need to make debugging easier?

**Comment by GoFroggyRun** [comment]
@martinholmer 
Thanks for the comment and advice.
I was actually thinking of something like `test.py` that could present all the variables to be inputed for observations in `puf`. The option you mentioned definitely will work, but having all the variables in the `puf` file written out might be time-consuming (you could consider that I want to have some parallel observations in `puf` that have been extrapolated). I have pseudocode that should work after uppercase/lowercase issue in #425 being fixed. And I was thinking of embedding such functionality inside `test.py`, as an alternative option of outputs. Does this make sense?

**Comment by martinholmer** [comment]
Sean Wang ( @GoFroggyRun ) said:
> I have pseudocode that should work after uppercase/lowercase issue in #425 being fixed. And I was thinking of embedding such functionality inside test.py, as an alternative option of outputs. Does this make sense?

I'm glad you're thinking about solutions.  I can't really say whether or not your idea is the best strategy until is see the pseudocode.  See that will give me a much clearer idea of the information you're seeking.

**Comment by martinholmer** [comment]
@GoFroggyRun,  Where do we stand on Tax-Calculator issue #507, which you raised on 15-Dec-2015?
Are there Tax-Calculator capabilities that are missing as of the current (14-Feb-2016) version?  If so, can you explain exactly what capabilities are missing and why they need to be added?   Thanks.

**Comment by martinholmer** [comment]
Pull request #597 resolves issue #507, so it is being closed.
